File: server/app/services/row_packer.py
from __future__ import annotations

import math
from array import array

from ..models.domain import DomainError
from .preparation_control import checked


def pack_footprints(items, clearance=4, max_index_bytes=128 * 1024 * 1024):
    if isinstance(clearance, bool) or not isinstance(clearance, (int, float)) or not math.isfinite(clearance) or clearance < 0:
        raise ValueError("Invalid packing clearance.")
    for item in checked(items):
        start, end = item["footprintStart"], item["footprintEnd"]
        if any(isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(value) for value in (start, end)) or end + clearance <= start:
            raise ValueError("Invalid packing footprint.")
    if len(items) < 128:
        tracks, rows = [], []
        for item in checked(items):
            start, end = item["footprintStart"], item["footprintEnd"]
            row = 0
            while row < len(tracks) and any(not (end + clearance <= left or start >= right + clearance) for left, right in tracks[row]):
                row += 1
            if row == len(tracks):
                tracks.append([])
            tracks[row].append((start, end))
            rows.append(row)
        return {"rows": rows, "count": len(tracks)}
    # Expanded half-open intervals encode the existing clearance rule exactly.
    points = sorted({coordinate for item in items for coordinate in (item["footprintStart"], item["footprintEnd"] + clearance)})
    coordinates = {value: index for index, value in enumerate(points)}
    size = len(points) - 1
    capacity = size * 4 + 4
    accounted_bytes = capacity * 24

    def require_capacity():
        if accounted_bytes > max_index_bytes:
            raise DomainError("layout_capacity", "Packing index capacity exceeded; narrow the time range or filters.", 413)

    require_capacity()
    aggregate, own = [0] * capacity, [0] * capacity
    aggregate_bits, own_bits = array("I", [0]) * capacity, array("I", [0]) * capacity

    def mark(values, widths, node, bit, row):
        nonlocal accounted_bytes
        previous, next_width = widths[node], row + 1
        if next_width > previous:
            accounted_bytes += (32 if previous == 0 else 0) + 8 * ((next_width + 63) // 64 - (previous + 63) // 64)
            require_capacity()
            widths[node] = next_width
        values[node] |= bit

    def occupied(node, left, right, start, end):
        if start <= left and right <= end:
            return aggregate[node]
        middle = (left + right) // 2
        result = own[node]
        if start < middle:
            result |= occupied(node * 2, left, middle, start, end)
        if end > middle:
            result |= occupied(node * 2 + 1, middle, right, start, end)
        return result

    def insert(node, left, right, start, end, bit, row):
        mark(aggregate, aggregate_bits, node, bit, row)
        if start <= left and right <= end:
            mark(own, own_bits, node, bit, row)
            return
        middle = (left + right) // 2
        if start < middle:
            insert(node * 2, left, middle, start, end, bit, row)
        if end > middle:
            insert(node * 2 + 1, middle, right, start, end, bit, row)

    rows, count, all_rows = [], 0, 0
    for item in checked(items):
        start, end = coordinates[item["footprintStart"]], coordinates[item["footprintEnd"] + clearance]
        blocked = occupied(1, 0, size, start, end)
        if blocked == all_rows:
            row = count
            count += 1
            bit = 1 << row
            all_rows |= bit
        else:
            bit = (blocked + 1) & ~blocked
            row = bit.bit_length() - 1
        rows.append(row)
        insert(1, 0, size, start, end, bit, row)
    return {"rows": rows, "count": count}

File: client/src/timeline/row-packer.js
export function packFootprints(items, clearance = 4, maxIndexBytes = 128 * 1024 * 1024) {
  if (!Number.isFinite(clearance) || clearance < 0 || items.some(item => !Number.isFinite(item.footprintStart) || !Number.isFinite(item.footprintEnd) || item.footprintEnd + clearance <= item.footprintStart)) throw new RangeError('Invalid packing footprint');
  if (items.length < 128) {
    const tracks = [], rows = [];
    for (const item of items) {
      let row = tracks.findIndex(intervals => intervals.every(([a, b]) => item.footprintEnd + clearance <= a || item.footprintStart >= b + clearance));
      if (row < 0) { row = tracks.length; tracks.push([]); }
      tracks[row].push([item.footprintStart, item.footprintEnd]); rows.push(row);
    }
    return { rows, count: tracks.length };
  }
  // Expanded half-open intervals exactly encode the existing clearance rule.
  // Range bitsets find the first available row without reordering any records.
  const points = [...new Set(items.flatMap(item => [item.footprintStart, item.footprintEnd + clearance]))].sort((a, b) => a - b);
  const coordinates = new Map(points.map((value, index) => [value, index])), size = points.length - 1;
  const capacity = size * 4 + 4;
  let accountedBytes = capacity * 24;
  const capacityError = () => Object.assign(new Error('Packing index capacity exceeded; narrow the time range or filters'), { code: 'layout_capacity', status: 413 });
  if (accountedBytes > maxIndexBytes) throw capacityError();
  const aggregate = new Array(capacity).fill(0n), own = new Array(capacity).fill(0n);
  const aggregateBits = new Uint32Array(capacity), ownBits = new Uint32Array(capacity);
  function mark(values, widths, node, bit, row) {
    const previous = widths[node], next = row + 1;
    if (next > previous) {
      accountedBytes += (previous === 0 ? 32 : 0) + 8 * (Math.ceil(next / 64) - Math.ceil(previous / 64));
      if (accountedBytes > maxIndexBytes) throw capacityError();
      widths[node] = next;
    }
    values[node] |= bit;
  }
  function occupied(node, left, right, start, end) {
    if (start <= left && right <= end) return aggregate[node];
    const middle = Math.floor((left + right) / 2);
    let result = own[node];
    if (start < middle) result |= occupied(node * 2, left, middle, start, end);
    if (end > middle) result |= occupied(node * 2 + 1, middle, right, start, end);
    return result;
  }
  function insert(node, left, right, start, end, bit, row) {
    mark(aggregate, aggregateBits, node, bit, row);
    if (start <= left && right <= end) { mark(own, ownBits, node, bit, row); return; }
    const middle = Math.floor((left + right) / 2);
    if (start < middle) insert(node * 2, left, middle, start, end, bit, row);
    if (end > middle) insert(node * 2 + 1, middle, right, start, end, bit, row);
  }
  const rows = []; let count = 0, allRows = 0n;
  for (const item of items) {
    const start = coordinates.get(item.footprintStart), end = coordinates.get(item.footprintEnd + clearance);
    const blocked = occupied(1, 0, size, start, end);
    let row, bit;
    if (blocked === allRows) { row = count++; bit = 1n << BigInt(row); allRows |= bit; }
    else { bit = (blocked + 1n) & ~blocked; const hex = bit.toString(16); row = (hex.length - 1) * 4 + Math.log2(Number.parseInt(hex[0], 16)); }
    rows.push(row); insert(1, 0, size, start, end, bit, row);
  }
  return { rows, count };
}
